File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-5.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int GoodNodes(TreeNode root) {
        int DFS(TreeNode root, int maxSeen) {
            if (root == null) return 0;

            int count = root.val >= maxSeen ? 1 : 0;
            maxSeen = Math.Max(maxSeen, root.val);

            return count + DFS(root.left, maxSeen) + DFS(root.right, maxSeen);
        }

        return DFS(root, root.val);
    }
}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-5.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if (preorder.empty() || inorder.empty()) return nullptr;

        TreeNode* root = new TreeNode(preorder[0]);
        int mid = find(inorder.begin(), inorder.end(), preorder[0]) - inorder.begin();

        vector<int> preorderLeft(preorder.begin() + 1, preorder.begin() + mid + 1);
        vector<int> inorderLeft(inorder.begin(), inorder.begin() + mid);
        vector<int> preorderRight(preorder.begin() + mid + 1, preorder.end());
        vector<int> inorderRight(inorder.begin() + mid + 1, inorder.end());

        root->left = buildTree(preorderLeft, inorderLeft);
        root->right = buildTree(preorderRight, inorderRight);

        return root;
    }
};
